test/lib/jdk/test/lib/cds/CDSAppTester.java

```diff
@@ -189,9 +192,37 @@ private OutputAnalyzer dumpStaticArchive() throws Exception {
         return executeAndCheck(cmdLine, runMode, staticArchiveFile, staticArchiveFileLog);
     }
 
+    // Creating a dynamic CDS archive (with -XX:ArchiveClassesAtExit=<foo>.jsa) requires that the current
+    // JVM process is using a static archive (which is usually the default CDS archive included in the JDK).
+    // However, if the JDK doesn't include a default CDS archive that's compatible with the set of
+    // VM options used by this test, we need to create a temporary static archive to be used with -XX:ArchiveClassesAtExit.
+    private String getBaseArchiveForDynamicArchive() throws Exception {
+        WhiteBox wb = WhiteBox.getWhiteBox();
+        if (wb.isSharingEnabled()) {
+            // This current JVM is able to use a default CDS archive included by the JDK, so
+            // if we launch a JVM child process (with the same set of options as the current JVM),
+            // that process is also able to use the same default CDS archive for creating
+            // a dynamic archive.
+            return null;
+        } else {
+            // This current JVM is unable to use a default CDS archive, so let's create a temporary
+            // static archive to be used with -XX:ArchiveClassesAtExit.
+            File f = new File(tempBaseArchiveFile);
+            if (!f.exists()) {
+                CDSOptions opts = new CDSOptions();
+                opts.setArchiveName(tempBaseArchiveFile);
+                opts.addSuffix("-Djava.class.path=");
+                OutputAnalyzer out = CDSTestUtils.createArchive(opts);
```

Comment:
Just to be clear, this creates the right kind of archive based on whether +UCOH is used or not?

Reply:
Yes, `CDSTestUtils.createArchive()` uses `ProcessTools.createTestJavaProcessBuilder()` to launch a child process to create the archive, so it passes the test vm options to the child.